src/Jaeger.Core/Samplers/RateLimitingSampler.cs

```diff
@@ -1,51 +1,54 @@
 using System;
 using System.Collections.Generic;
+using System.Collections.ObjectModel;
 using Jaeger.Core.Util;
 
 namespace Jaeger.Core.Samplers
 {
     /// <summary>
-    /// Only used for unit testing
+    /// <see cref="RateLimitingSampler"/> creates a sampler that samples at most maxTracesPerSecond. The distribution of sampled
+    /// traces follows burstiness of the service, i.e. a service with uniformly distributed requests will have those
+    /// requests sampled uniformly as well, but if requests are bursty, especially sub-second, then a number of
+    /// sequential requests can be sampled each second.
     /// </summary>
-    internal interface IRateLimitingSampler : ISampler
+    public class RateLimitingSampler : ValueObject, ISampler
     {
-        double MaxTracesPerSecond { get; }
-    }
+        public const string Type = "ratelimiting";
 
-    // RateLimitingSampler creates a sampler that samples at most maxTracesPerSecond. The distribution of sampled
-    // traces follows burstiness of the service, i.e. a service with uniformly distributed requests will have those
-    // requests sampled uniformly as well, but if requests are bursty, especially sub-second, then a number of
-    // sequential requests can be sampled each second.
-    public class RateLimitingSampler : IRateLimitingSampler
-    {
-        internal readonly IRateLimiter _rateLimiter;
-        private readonly Dictionary<string, object> _tags;
+        private readonly RateLimiter _rateLimiter;
+        private readonly ReadOnlyDictionary<string, object> _tags;
 
         public double MaxTracesPerSecond { get; }
 
         public RateLimitingSampler(double maxTracesPerSecond)
-            : this(maxTracesPerSecond, new RateLimiter(maxTracesPerSecond, Math.Max(maxTracesPerSecond, 1.0)))
-        {}
-
-        public RateLimitingSampler(double maxTracesPerSecond, IRateLimiter rateLimiter)
         {
             MaxTracesPerSecond = maxTracesPerSecond;
+            _rateLimiter = new RateLimiter(maxTracesPerSecond, Math.Max(maxTracesPerSecond, 1.0));
+
+            _tags = new ReadOnlyDictionary<string, object>(new Dictionary<string, object> {
```

Comment:
Hmm. Having the strong guarantee with the extra `ReadOnlyDictionary` object might not be necessary though... If we just use the regular `Dictionary`, as you've suggested, someone would have to cast the `IReadOnlyDictionary` back to a `Dictionary` to manipulate it. That's not really a scenario we need to protect us from in this library so we might as well save some memory and return the objects themselves.

Reply:
I assume you already saw that the `IReadOnlyDictionary` interface does not have an `Add` method and that the throwing code in the `ReadOnlyDictionary` class is for fulfilling the `IDictionary` interface. So it's all just about saving memory and runtime :) I saw that you already commited it, I just wanted to state it here again in case someone reads it up.